#**Maestría en Inteligencia Artificial Aplicada**
##**Curso: Procesamiento de Lenguaje Natural (NLP)**
###Tecnológico de Monterrey
###Prof Luis Eduardo Falcón Morales

## **Adtividad de la Semana - regex**
###**Introducción a la limpieza y procesamiento de texto.**

* **Nombre:** Gustavo Naranjo Hernandez

* **Matrícula:** A00814322

En esta actividad deberás utilizar el archivo llamado yelp_labelled.txt

El archivo contiene comentarios en inglés sobre servicios de comida de la página de Yelp: https://www.yelp.com/ .

Son mil comentarios y forman parte del conjunto de datos que se encuentra en el Machine Learning Repository de la UCI, llamado "Sentiment Labelled Sentences": https://archive.ics.uci.edu/ml/datasets/Sentiment+Labelled+Sentences#


#**Parte 1. Cargamos los datos.**   

Cargar los datos del archivo indicado y obtener una lista de longitud de 1000 strings/comentarios.

Por el momento solamente requerimos las bibliotecas de Numpy y re, para el manejo de los arreglos y de las expresiones regulares en Python.

En particular, no necesitarás en esta actividad la biblioteca de Pandas.

###**NOTA: En esta actividad no debes importar nada más, con estas dos bibliotecas será *suficiente*.**

In [1]:
import numpy as np    # importamos Numpy para el manejo de los arreglos.
import re             # importamos re para el manejo de las expresiones regulares.

In [5]:
# Ejecuta las siguientes instrucciones para cargar la información del achivo dado:

with open('yelp_labelled.txt',        # puedes actualizar la ruta a tu archivo, en dado caso.
          mode='r',     # abrimos el archivo en modo lectura.
          ) as f:
    docs = f.readlines()    # separamos cada comentario por líneas

f.close()  # ya que tenemos la información en la variable docs, cerramos el archivo

In [6]:
type(docs) == list   # Verifica que tu variable "docs" es una lista

True

In [7]:
len(docs)==1000  # verifica que la longitud de "docs" es de mil comentarios.

True

In [8]:
docs[0:10]     # observa algunos de los primeros comentarios

['Wow... Loved this place.\t1\n',
 'Crust is not good.\t0\n',
 'Not tasty and the texture was just nasty.\t0\n',
 'Stopped by during the late May bank holiday off Rick Steve recommendation and loved it.\t1\n',
 'The selection on the menu was great and so were the prices.\t1\n',
 'Now I am getting angry and I want my damn pho.\t0\n',
 "Honeslty it didn't taste THAT fresh.)\t0\n",
 'The potatoes were like rubber and you could tell they had been made up ahead of time being kept under a warmer.\t0\n',
 'The fries were great too.\t1\n',
 'A great touch.\t1\n']

#**Parte 2: sección de preguntas (regex).**   


##**Instrucciones:**

###**A continuación deberás contestar cada una de las preguntas que te piden usando expresiones regulares (regex).**

###**Por el momento no hay restricción en cuanto al número de líneas de código que agregues, pero trata de incluir las mínimas posibles.**

*   **Pregunta 1.**

Busca y elimina todos los saltos de línea '\n', así como las etiquetas de 0s y 1s que se encuentran al final de cada comentario. En esta actividad solo trabajaremos con la limpieza de datos, por lo que no requeriremos dichas etiquetas por el momento.

Una vez finalizado, imprime los primeros 10 comentarios del resultado obtenido.


In [9]:
# Pregunta 1: Eliminamos los saltos de línea y las etiquetas (0 o 1) al final de cada comentario
docs_clean = [re.sub(r'\t[01]\n?$', '', doc) for doc in docs]

# Verificamos los primeros 10 comentarios
print(docs_clean[0:10])

['Wow... Loved this place.', 'Crust is not good.', 'Not tasty and the texture was just nasty.', 'Stopped by during the late May bank holiday off Rick Steve recommendation and loved it.', 'The selection on the menu was great and so were the prices.', 'Now I am getting angry and I want my damn pho.', "Honeslty it didn't taste THAT fresh.)", 'The potatoes were like rubber and you could tell they had been made up ahead of time being kept under a warmer.', 'The fries were great too.', 'A great touch.']


In [10]:
# Verificamos que seguimos teniendo 1000 comentarios
print(f'Total de comentarios: {len(docs_clean)}')

Total de comentarios: 1000


**NOTA: A partir de ahora, deberás trabajar con los comentarios obtenidos de esta primer pregunta.**

*   **Pregunta 2.**  

Busca e imprime todas las palabras que terminan con dos o más signos de admiración seguidos, por ejemplo "!!!".

Debes imprimir tanto la palabra como la totalidad de signos de admiración que le siguen.

Indica cuántos resultados obtuviste.



In [33]:
# Pregunta 2: Palabras que terminan con 2 o más signos de admiración seguidos
p2_resultados = []
for doc in docs_clean:
    matches = re.findall(r'\w+!{2,}', doc)
    p2_resultados.extend(matches)

print('Palabras encontradas:')
print(p2_resultados)

Palabras encontradas:
['Firehouse!!!!!', 'APPETIZERS!!!', 'amazing!!!', 'buffet!!!', 'good!!', 'it!!!!', 'DELICIOUS!!', 'amazing!!', 'shawarrrrrrma!!!!!!', 'yucky!!!', 'steak!!!!!', 'delicious!!!', 'far!!', 'biscuits!!!', 'dry!!', 'disappointing!!!', 'awesome!!', 'Up!!', 'FLY!!!!!!!!', 'here!!!', 'great!!!!!!!!!!!!!!', 'packed!!', 'otherwise!!', 'amazing!!!!!!!!!!!!!!!!!!!', 'style!!', 'disappointed!!']


In [34]:
print(f'Número de resultados: {len(p2_resultados)}')

Número de resultados: 26


*   **Pregunta 3.**  

Busca e imprime todas las palabras o cadena de 2 o más catacteres que están escritas totalmente en mayúsculas. Cada coincidencia debe ser una sola palabra (cadena de caracteres).

Indica cuántas palabras encontraste.



In [35]:
# Pregunta 3: Palabras de 2+ caracteres escritas totalmente en mayúsculas
p3_resultados = []
for doc in docs_clean:
    matches = re.findall(r'\b[A-Z]{2,}\b', doc)
    p3_resultados.extend(matches)

print('Palabras en mayúsculas encontradas:')
print(p3_resultados)

Palabras en mayúsculas encontradas:
['THAT', 'APPETIZERS', 'WILL', 'NEVER', 'EVER', 'STEP', 'FORWARD', 'IN', 'IT', 'AGAIN', 'LOVED', 'AND', 'REAL', 'BITCHES', 'NYC', 'STALE', 'DELICIOUS', 'WORST', 'EXPERIENCE', 'EVER', 'ALL', 'BARGAIN', 'TV', 'NONE', 'FREEZING', 'AYCE', 'FLAVOR', 'NEVER', 'BBQ', 'UNREAL', 'OMG', 'BETTER', 'BLAND', 'RUDE', 'INCONSIDERATE', 'MANAGEMENT', 'WILL', 'NEVER', 'EVER', 'GO', 'BACK', 'AND', 'HAVE', 'TOLD', 'MANY', 'PEOPLE', 'WHAT', 'HAD', 'HAPPENED', 'TOTAL', 'WASTE', 'OF', 'TIME', 'FS', 'AZ', 'LOVED', 'CONCLUSION', 'BEST', 'GO', 'NOW', 'GC', 'AVOID', 'THIS', 'ESTABLISHMENT', 'AN', 'HOUR', 'NASTY', 'OMG', 'NO', 'BEST', 'THE', 'OWNERS', 'REALLY', 'REALLY', 'PERFECT', 'SCREAMS', 'LEGIT', 'MGM', 'BEST', 'FLY', 'FLY', 'FANTASTIC', 'GREAT', 'OK', 'WAY', 'MUST', 'HAVE', 'OK', 'OVERPRICED', 'BARE', 'HANDS', 'WEAK', 'SHOULD', 'RI', 'VERY', 'NOT']


In [36]:
print(f'Número de palabras en mayúsculas: {len(p3_resultados)}')

Número de palabras en mayúsculas: 96


*   **Pregunta 4.**  

Busca e imprime los comentarios en donde todos los caracteres alfabéticos (letras) están en mayúsculas.

Cada coincidencia encontrada debe ser todo el comentario/enunciado.

Indica cuántos resultados obtuviste.


In [37]:
# Pregunta 4: Comentarios donde TODOS los caracteres alfabéticos están en mayúsculas
p4_resultados = [doc for doc in docs_clean if re.fullmatch(r'[^a-z]+', doc)]

print('Comentarios con todos los alfabéticos en mayúsculas:')
for d in p4_resultados:
    print(d)

Comentarios con todos los alfabéticos en mayúsculas:
DELICIOUS!!
RUDE & INCONSIDERATE MANAGEMENT.
WILL NEVER EVER GO BACK AND HAVE TOLD MANY PEOPLE WHAT HAD HAPPENED.
TOTAL WASTE OF TIME.
AVOID THIS ESTABLISHMENT!


In [38]:
print(f'Número de resultados: {len(p4_resultados)}')

Número de resultados: 5


*   **Pregunta 5.**  

Busca e imprime todas las palabras que tengan una vocal acentuada, del tipo á, é, í, ó, ú.

Indica cuántos resultados obtuviste.

In [39]:
# Pregunta 5: Palabras con vocal acentuada (á, é, í, ó, ú)
p5_resultados = []
for doc in docs_clean:
    matches = re.findall(r'\b\w*[áéíóú]\w*\b', doc, re.IGNORECASE)
    p5_resultados.extend(matches)

print('Palabras con vocal acentuada:')
print(p5_resultados)

Palabras con vocal acentuada:
['fiancé', 'Café', 'puréed']


In [40]:
print(f'Número de resultados: {len(p5_resultados)}')

Número de resultados: 3


*   **Pregunta 6.**  

Busca e imprime todas las cantidades numéricas monetarias, enteras o con decimales, que inician con el símbolo $\$$.

Indica cuántos resultados obtuviste.

In [19]:
# Pregunta 6: Cantidades monetarias que inician con $
p6_resultados = []
for doc in docs_clean:
    matches = re.findall(r'\$\d+(?:\.\d+)?', doc)
    p6_resultados.extend(matches)

print('Cantidades monetarias encontradas:')
print(p6_resultados)

Cantidades monetarias encontradas:
['$20', '$4.00', '$17', '$3', '$35', '$7.85', '$12', '$11.99']


In [20]:
print(f'Número de resultados: {len(p6_resultados)}')

Número de resultados: 8


*   **Pregunta 7.**  

Busca e imprime todas las palabras que sean variantes de la palabra "love" (amor). No deberá importar si incluyen mayúsculas o minúsculas, o la manera en que esté conjugada o alguna otra variación que se haga con dicha palabra, como repetir letras por ejemplo.

Observa que deben ser variantes de la palabra love (amor) y no variantes de dicha cadena de caracteres.

Indica cuántos resultados obtuviste.

In [21]:
# Pregunta 7: Variantes de la palabra 'love'
# Incluye: love, loved, loves, lover, lovely, LOVE, LOVED, loooove, etc.
p7_resultados = []
for doc in docs_clean:
    matches = re.findall(r'\b[lL]+[oO]+[vV]+[eE]+\w*\b', doc)
    p7_resultados.extend(matches)

print('Variantes de "love" encontradas:')
print(p7_resultados)

Variantes de "love" encontradas:
['Loved', 'loved', 'Loved', 'love', 'loves', 'LOVED', 'lovers', 'love', 'lovers', 'Love', 'loved', 'loved', 'love', 'love', 'love', 'loved', 'love', 'loved', 'Love', 'LOVED', 'love', 'lovely', 'love', 'lovely', 'love', 'lover', 'loved', 'love', 'love', 'love', 'love', 'love', 'love', 'love', 'love']


In [22]:
print(f'Número de resultados: {len(p7_resultados)}')

Número de resultados: 35


*   **Pregunta 8.**  

Busca e imprime todas las palabras que sean variantes de las palabras en inglés "so" y "good", y que tengan dos o más "o" en "so" y 3 o más "o" en good.

Observa que solo debes incluir variantes de estas palabras en inglés y no de dichas cadenas de caracteres.

Indica además cuántas encontraste.


In [23]:
# Pregunta 8: Variantes de 'so' (2+ o's) y 'good' (3+ o's)
p8_so = []
p8_good = []
for doc in docs_clean:
    matches_so = re.findall(r'\bso{2,}\b', doc, re.IGNORECASE)
    matches_good = re.findall(r'\bgo{3,}d\b', doc, re.IGNORECASE)
    p8_so.extend(matches_so)
    p8_good.extend(matches_good)

print(f'Variantes de "so" (2+ o): {p8_so}')
print(f'Variantes de "good" (3+ o): {p8_good}')

Variantes de "so" (2+ o): ['Sooooo', 'soooo', 'soooooo', 'soooo']
Variantes de "good" (3+ o): []


In [24]:
print(f'Resultados "so": {len(p8_so)}')
print(f'Resultados "good": {len(p8_good)}')
print(f'Total: {len(p8_so) + len(p8_good)}')

Resultados "so": 4
Resultados "good": 0
Total: 4


*   **Pregunta 9.**  

Busca e imprime todas las palabras que tengan una longitud mayor estrictamente a 12 caracteres alfabéticos.

No se consideran los signos de puntuación o caracteres especiales en la longitud de estas cadenas, solo caracteres alfabéticos en mayúsculas o minúsculas.

Indica la cantidad de palabras encontradas.


In [25]:
# Pregunta 9: Palabras con más de 12 caracteres alfabéticos (sin contar puntuación)
p9_resultados = []
for doc in docs_clean:
    words = re.findall(r'\b\w+\b', doc)
    for w in words:
        alpha_only = re.sub(r'[^a-zA-Z]', '', w)
        if len(alpha_only) > 12:
            p9_resultados.append(w)

print('Palabras con más de 12 caracteres alfabéticos:')
print(p9_resultados)

Palabras con más de 12 caracteres alfabéticos:
['recommendation', 'establishment', 'underwhelming', 'disappointing', 'disappointing', 'Unfortunately', 'accommodations', 'recommendation', 'shawarrrrrrma', 'INCONSIDERATE', 'disappointment', 'establishment', 'extraordinary', 'ESTABLISHMENT', 'disappointing', 'recommendation', 'establishment', 'Unfortunately', 'Wienerschnitzel', 'unfortunately', 'Mediterranean', 'unprofessional', 'disappointing', 'unexperienced', 'Disappointing', 'WAAAAAAyyyyyyyyyy', 'disappointment', 'disappointment', 'deuchebaggery', 'disappointment', 'Unfortunately', 'disapppointment', 'circumstances', 'underwhelming']


In [26]:
print(f'Número de resultados: {len(p9_resultados)}')

Número de resultados: 34


*   **Pregunta 10.**  

Busca e imprime todas las palabras (caracteres alfabéticos) que inician con una letra mayúscula y terminan con una minúscula, pero donde además dicha palabra no sea la primera palabra del comentario (string).

Indica la cantidad de resultados obtenidos.

In [27]:
# Pregunta 10: Palabras (no primera del comentario) que inician con mayúscula y terminan en minúscula
p10_resultados = []
for doc in docs_clean:
    # Eliminar la primera palabra del comentario
    resto = re.sub(r'^\S+\s*', '', doc)
    matches = re.findall(r'\b[A-Z][a-zA-Z]*[a-z]\b', resto)
    p10_resultados.extend(matches)

print('Primeros 20 resultados:')
print(p10_resultados[:20])

Primeros 20 resultados:
['Loved', 'May', 'Rick', 'Steve', 'Cape', 'Cod', 'Vegas', 'Burrittos', 'Blah', 'The', 'They', 'Mexican', 'Luke', 'Our', 'Hiro', 'Firehouse', 'Greek', 'Greek', 'Heart', 'Attack']


In [28]:
print(f'Número de resultados: {len(p10_resultados)}')

Número de resultados: 271


*   **Pregunta 11.**  

Busca e imprime la secuencia de dos o más palabras que están separadas por un guion, "-", sin que tengan espacios en blanco entre ellas.

Por ejemplo "Go-Kart" sería válido, pero "Go  -Kart" o "Go  -  Kart" no lo serían, ya que incluyen algún espacio en blanco.

Indica la cantidad de resultados obtenidos.

In [29]:
# Pregunta 11: Secuencias de 2+ palabras separadas por guion sin espacios
p11_resultados = []
for doc in docs_clean:
    matches = re.findall(r'\b\w+(?:-\w+)+\b', doc)
    p11_resultados.extend(matches)

print('Secuencias con guion encontradas:')
print(p11_resultados)

Secuencias con guion encontradas:
['flat-lined', 'hands-down', 'must-stop', 'sub-par', 'Service-check', 'in-house', 'been-stepped-in-and-tracked-everywhere', 'multi-grain', 'to-go', 'non-customer', 'High-quality', 'sit-down', 'over-whelm', 'low-key', 'non-fancy', 'golden-crispy', 'over-priced', 'over-hip', 'under-services']


In [30]:
print(f'Número de resultados: {len(p11_resultados)}')

Número de resultados: 19


*   **Pregunta 12.**  

Busca e imprime todas las palabras que terminan con la conjugación en inglés "ing" o "ed".

No deben importar mayúsculas o minúsculas.

Indica la cantidad de palabras que encontraste de cada una.

In [31]:
# Pregunta 12: Palabras que terminan en 'ing' o 'ed'
p12_ing = []
p12_ed = []
for doc in docs_clean:
    matches_ing = re.findall(r'\b\w+ing\b', doc, re.IGNORECASE)
    matches_ed  = re.findall(r'\b\w+ed\b',  doc, re.IGNORECASE)
    p12_ing.extend(matches_ing)
    p12_ed.extend(matches_ed)

print(f'Palabras que terminan en "ing": {len(p12_ing)}')
print(f'Palabras que terminan en "ed":  {len(p12_ed)}')

Palabras que terminan en "ing": 280
Palabras que terminan en "ed":  339


In [32]:
print('Muestra de palabras con "ing":', p12_ing[:10])
print('Muestra de palabras con "ed":', p12_ed[:10])

Muestra de palabras con "ing": ['during', 'getting', 'being', 'being', 'amazing', 'running', 'redeeming', 'getting', 'thing', 'dressing']
Muestra de palabras con "ed": ['Loved', 'Stopped', 'loved', 'ended', 'overpriced', 'tried', 'disgusted', 'shocked', 'recommended', 'performed']


#**Parte 3. Proceso de limpieza.**

*   **Pregunta 13.**  

Ahora realiza un proceso de limpieza del corpus que incluya los siguientes procesos:

*   Solo se deben considerar caracteres alfabéticos, incluyendo algún tipo de acentuación. Es decir, se eliminan todos los signos de puntuación y caracteres especiales.
*   Todos los caracteres alfabéticos se transforman a minúsculas.
*   Se deben eliminar todos los espacios en blanco adicionales que se puedan encontrar en cada comentario.

Al finalizar dicho proceso de limpieza, imprime el resultado de los primeros 10 comentarios resultantes.
   




In [41]:
# Pregunta 13: Proceso de limpieza del corpus
def limpiar_texto(texto):
    # 1. Conservar solo caracteres alfabéticos (incluyendo acentuadas) y espacios
    texto = re.sub(r'[^a-zA-ZáéíóúÁÉÍÓÚ\s]', ' ', texto)
    # 2. Convertir todo a minúsculas
    texto = texto.lower()
    # 3. Eliminar espacios en blanco adicionales
    texto = re.sub(r'\s+', ' ', texto).strip()
    return texto

docs_limpios = [limpiar_texto(doc) for doc in docs_clean]

In [42]:
print('Primeros 10 comentarios limpios:')
for d in docs_limpios[:10]:
    print(d)

Primeros 10 comentarios limpios:
wow loved this place
crust is not good
not tasty and the texture was just nasty
stopped by during the late may bank holiday off rick steve recommendation and loved it
the selection on the menu was great and so were the prices
now i am getting angry and i want my damn pho
honeslty it didn t taste that fresh
the potatoes were like rubber and you could tell they had been made up ahead of time being kept under a warmer
the fries were great too
a great touch


*   **Pregunta 14.**  

Supongamos que en un primer análisis básico consideramos como nuestro nuevo corpus el texto obtenido después del proceso de limpieza del ejercicio anterior.  

A partir de este nuevo corpus consideraremos cada palabra (cadena de caracteres) resultante como los "tokens" de nuestro estudio.

Y a partir de este documento tokenizado generamos nuestro diccionario de tokens.

¿Cuántos tokens diferentes tiene nuestro diccionario?


In [43]:
# Pregunta 14: Construir el diccionario de tokens únicos
todos_los_tokens = []
for doc in docs_limpios:
    tokens = doc.split()
    todos_los_tokens.extend(tokens)

diccionario = set(todos_los_tokens)
print(f'Total de tokens (con repetición): {len(todos_los_tokens)}')

Total de tokens (con repetición): 11039


In [44]:
print(f'Tokens únicos en el diccionario: {len(diccionario)}')

Tokens únicos en el diccionario: 2021


*   **Pregunta 15.**  

Finalmente, en este ejercicio definiremos nuestro conjunto de palabras denominado "stopwords", es decir, las palabras (cadenas de caracteres) que tienen frecuencias muy altas, pero que no aportan mucho valor para el significado/entendimiento de cada comentario.

Recuerda que ejemplos de stopwords son artículos, adverbios, conectivos, etcétera.

* **15a.** Con base a la lista de stopwords que se te proporciona, realiza un proceso de limpieza eliminando todas estas palabras del corpus obtenido en el ejercicio anterior. Imprime los primeros 5 comentarios resultantes.

* **15b.** ¿De qué tamaño quedó tu documento? Es decir, ¿cuántos tokens tiene finalmente todo tu nuevo documento de comentarios procesado?

* **15c.** ¿Cuántos tokens diferentes quedaron? Es decir, ¿cuántos tokens únicos tiene tu vocabulario final?

In [45]:
# Debes considerar la siguiente lista como tu conjunto de stopwords:
mis_stopwords = ['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', 'your', 'yours', 'he', 'him', 'his', 'himself', 'she', 'her', 'hers', 'herself', 'it', 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'only', 'own', 'same', 'so', 'than', 'too', 'very', 's', 't', 'can', 'will', 'just', 'should', 'now', 'll']

In [46]:
# Pregunta 15a: Eliminar stopwords del corpus
stopwords_set = set(mis_stopwords)

docs_sin_stopwords = []
for doc in docs_limpios:
    tokens = doc.split()
    filtrados = [t for t in tokens if t not in stopwords_set]
    docs_sin_stopwords.append(' '.join(filtrados))

print('Primeros 5 comentarios tras eliminar stopwords:')
for d in docs_sin_stopwords[:5]:
    print(d)

Primeros 5 comentarios tras eliminar stopwords:
wow loved place
crust not good
not tasty texture nasty
stopped late may bank holiday off rick steve recommendation loved
selection menu great prices


In [47]:
# 15b: Total de tokens tras eliminar stopwords
tokens_finales = []
for doc in docs_sin_stopwords:
    tokens_finales.extend(doc.split())

print(f'15b - Total de tokens en el documento final: {len(tokens_finales)}')

# 15c: Tokens únicos
print(f'15c - Tokens únicos (vocabulario final): {len(set(tokens_finales))}')

15b - Total de tokens en el documento final: 5783
15c - Tokens únicos (vocabulario final): 1910


*   **Pregunta 15. Comentarios**

Incluye finalmente tus comentarios de la actividad.

En esta actividad aprendimos a usar expresiones regulares (`re`) para explorar y limpiar texto en Python. Fue especialmente interesante observar cómo los usuarios de Yelp escriben de manera informal: palabras en mayúsculas para enfatizar emociones, signos de exclamación repetidos, elongaciones de palabras como 'sooooo' o 'shawarrrrrrma', y uso de guiones para compuestos. El proceso de limpieza (Parte 3) redujo el vocabulario de 2,021 tokens únicos a 1,910 después de eliminar stopwords, y el total de tokens pasó de 11,039 a 5,783, lo que representa una reducción significativa del ruido textual antes de aplicar modelos de NLP.

##**Fin de la Actividad de la semana 2.**